test/monitors/whamm/monitors/generators/gen_hotness-mon/GenHotnessMon.v3

```diff
@@ -0,0 +1,247 @@
+                continue;
+            }
+        }
+        gen_probe(fd, mnemonic);
+    }
+
+    alloc_func(fd);
+    count_func(fd);
+    flush_func(fd);
+
+    module_end(fd);
+	System.fileClose(fd);
+	return 0;
+}
+
+def gen_probe(fd: int, opcode_name: string) {
+    func_start(fd, "(export \"wasm:opcode:", false);
+    write(fd, opcode_name);
+    write(fd, "($alloc(fid, pc))\") (param $entry i32)\n");
+
+    writet(fd, 2, "(call $count_probe (local.get $entry))");
+
+    func_end(fd);
+}
+
+def write_docs(fd: int) {
+    write(fd, ";; === Whamm Hotness Monitor ===\n");
+    write(fd, ";; Instrument EVERY location in the application and track each\n");
+    write(fd, ";; time it is executed dynamically with an in-memory counter.\n\n");
+    write(fd, ";; --> Output format: <fid>, pc=<pc>, [<counter>]\n\n");
+}
+
+def setup(fd: int) {
+    // imports
+    writet(fd, 1, "(import \"wizeng\" \"puti\" (func $puti (param i32)))\n");
+    writet(fd, 1, "(import \"wizeng\" \"puts\" (func $puts (param i32 i32)))\n\n");
+    
+    // memory/globals
+    writet(fd, 1, "(memory (export \"mem\") 2)   ;; no expansion checks\n");
+    writet(fd, 1, "(global $last_entry (mut i32) (i32.const 0))\n\n");
+
+    // data(data (i32.const 0xc00) "func=")
+    writet(fd, 1, "(data (i32.const 0xd00) \", pc=\")\n");
+    writet(fd, 1, "(data (i32.const 0xe00) \"\\n\")\n");
+    writet(fd, 1, "(data (i32.const 0xf00) \", [\")\n");
+    writet(fd, 1, "(data (i32.const 0xf10) \",\")\n");
+    writet(fd, 1, "(data (i32.const 0xf20) \"]\")\n\n");
+}
+
+def alloc_func(fd: int) {
+    func_start(fd, "(export \"$alloc\") (param $func i32) (param $pc i32) (result i32)", true);
+    writet(fd, 2, "(local $entry i32)\n\n");
+
+    writet(fd, 2, "global.get $last_entry\n");
+    writet(fd, 2, "local.set $entry\n\n");
+
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "local.get $func\n");
+    writet(fd, 2, "i32.store\n\n");
+
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "local.get $pc\n");
+    writet(fd, 2, "i32.store offset=4\n\n");
+
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "i32.const 1\n");
+    writet(fd, 2, "i32.store offset=8\n\n");
+
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "i32.const 20\n");
+    writet(fd, 2, "i32.add\n");
+    writet(fd, 2, "global.set $last_entry\n\n");
+
+    writet(fd, 2, "local.get $entry\n");
+
+    func_end(fd);
+}
+
+def count_func(fd: int) {
+    func_start(fd, "$count_probe (param $entry i32)", true);
+
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "local.get $entry\n");
+    writet(fd, 2, "i64.load offset=12  ;; count number of times fired\n");
+    writet(fd, 2, "i64.const 1\n");
+    writet(fd, 2, "i64.add\n");
+    writet(fd, 2, "i64.store offset=12\n");
+
+    func_end(fd);
+}
+
+def flush_func(fd: int) {
+    func_start(fd, "$flush (export \"wasm:exit\")", true);
+    
+    writet(fd, 2, "(local $entry i32)\n");
```

Comment:
There are lots of options for making this nicer, e.g. using a partially-applied function:

```
  var w = writet(fd, 2, _);
  w("line 1");
  w("line 2");
  w("line 3");
  w("line 4");
  w = writet(fd, 3, _);
  w("line 5");
  w("line 6");
  w("line 7");
```

You can make this a little nicer with a helper class that has the indentation state in it and use the "return this" pattern, like so:

```
var o = IndentingOutputter.new(); // choose your favorite name here

o.indent()
  .w("level 1 stuff")
  .w("level 1 stuff")
  .w("level 1 stuff")
  .indent()
    .w("level 2 stuff")
    .unindent()
```

Or other options are to take an array of strings and you can pass a literal:

```
 o.w([
  "line 1",
  "line 1",
  "line 1"
  ])

I usually just do this to get rid of as much noise as possible. Since the generator's performance doesn't matter at all, you can make it nice.

```